### DISCUSSION TITLE
Using RenderingIntent embedded in the profile (varing `intent` based on ICC metadata)

### DISCUSSION THREAD

**Discussion by paperboyo** [libvips/libvips#3474] [Help wanted]
Hello. Thank you so much for your work on VIPS!

Short version: is there a way for `vips_icc_transform` (and `vips_icc_import`?) to use `intent` present in the profile embedded in the source image instead of me having to declare it explicitly (or use default `relative`)?

Problem: `relative` works well as a default rendering intent. We would like to use VIPS to perform conversions from any CMYK profile into RGB. Most CMYK images were created from RGB data using `relative` rendering intent.
But at least for one, very common, [standard](http://www.ifra.com/WebSite/ifra.nsf/html/CONT_ISO_DOWNLOADS), CMYK profile, all conversions were performed and are still being performed using `perceptual` rendering intent (for a good reason).
Conversions were performed using Adobe Photoshop. When Photoshop converts from RGB to CMYK and user embeds the output profile in an image, Photoshop writes that embedded CMYK profile with a rendering intent **that was actually used for the conversion** ([6.1.11 Rendering Intent](https://www.color.org/icc32.pdf#page=28&zoom=auto,-443,523)).

So:
- if I convert sRGB into ISONewspaperv4 selecting Perceptual and save an image embedding the profile, the resulting image will have ISONewspaperv4 profile embedded that specifies `perceptual` as its rendering intent.
- if I convert sRGB into ISONewspaperv4 selecting Relative and save an image embedding the profile, the resulting image will have ISONewspaperv4 profile embedded that specifies `relative` as its rendering intent.

As converting those images back to RGB must use the same rendering intent used to create them in the first place, I would like to be able to set `intent` for `vips_icc_transform` to something like `icc_embedded`, so that the `intent` would depend on the rendering intent specified inside the profile embedded inside the image in question and could differ per-image.

I attach two images created from the same RGB data, converted to ISONewspaperv4 CMYK profile using perceptual and relative rendering intents.

For the image converted using Perceptual, exiftool reports:
`[ICC_Profile] Rendering Intent: Perceptual`

For the image converted using Relative, exiftool reports:
`[ICC_Profile] Rendering Intent: Media-Relative Colorimetric`

[Photoshop RGBtoCMYK conversions using different Rendering Intents.zip](https://github.com/libvips/libvips/files/11361315/Photoshop.RGBtoCMYK.conversions.using.different.Rendering.Intents.zip)

Is there a way for me to achieve this using VIPS alone? Can VIPS vary rendering intent based on ICC metadata? Thank you!

**Comment by jcupitt** [comment]
Hi @paperboyo,

That's interesting, I didn't realize pshop made and attached a custom profile.

I don't think there's such a thing as a default rendering intent tag in a profile though, is there? For a typical profile, libvips sees eg.:

```
description: sRGB IEC61966-2.1
manufacturer: IEC http://www.iec.ch
model: IEC 61966-2.1 Default RGB colour space - sRGB
copyright: Copyright (c) 1998 Hewlett-Packard Company
profile class: 0x6d6e7472
PCS: 0x58595a20 (XYZ )
matrix shaper: 1
version: 2.1
n_intents = 10
  0: Perceptual, in CLUT = 0, out CLUT = 0
  0x1: Relative colorimetric, in CLUT = 0, out CLUT = 0
  0x2: Saturation, in CLUT = 0, out CLUT = 0
  0x3: Absolute colorimetric, in CLUT = 0, out CLUT = 0
  0xa: Perceptual preserving black ink, in CLUT = 0, out CLUT = 0
  0xb: Relative colorimetric preserving black ink, in CLUT = 0, out CLUT = 0
  0xc: Saturation preserving black ink, in CLUT = 0, out CLUT = 0
  0xd: Perceptual preserving black plane, in CLUT = 0, out CLUT = 0
  0xe: Relative colorimetric preserving black plane, in CLUT = 0, out CLUT = 0
  0xf: Saturation preserving black plane, in CLUT = 0, out CLUT = 0
```

Is pshop making a custom profile with exactly one intent? In which case libvips may already do what you want -- if you ask for an intent that's not supported, it should default to the best available intent, which would be the one that's there (I think).

Do you have a sample image that's been written with one of these profiles?

**Comment by jcupitt** [comment]
I tried with your profile and I see:

```
$ vips icc_import ~/pics/cmyk2.jpg x.v --input-profile ~/pics/ISOnewspaper26v4.icc
loading input profile, intent relative, from blob 0x5571162cdd70
icc profile loaded from blob to make: 0x5571162cc110
description: ISOnewspaper26v4
copyright: Copyright by LOGO GmbH, Steinfurt
profile class: 0x70727472
PCS: 0x4c616220 (Lab )
matrix shaper: 0
version: 2.4
n_intents = 10
  0: Perceptual, in CLUT = 1, out CLUT = 1
  0x1: Relative colorimetric, in CLUT = 1, out CLUT = 1
  0x2: Saturation, in CLUT = 1, out CLUT = 1
  0x3: Absolute colorimetric, in CLUT = 1, out CLUT = 1
  0xa: Perceptual preserving black ink, in CLUT = 0, out CLUT = 0
  0xb: Relative colorimetric preserving black ink, in CLUT = 0, out CLUT = 0
  0xc: Saturation preserving black ink, in CLUT = 0, out CLUT = 0
  0xd: Perceptual preserving black plane, in CLUT = 0, out CLUT = 0
  0xe: Relative colorimetric preserving black plane, in CLUT = 0, out CLUT = 0
  0xf: Saturation preserving black plane, in CLUT = 0, out CLUT = 0
```

So it looks lie a pretty standard table profile with the usual four intents.

**Comment by paperboyo** [comment]
Hi, @jcupitt. Thanks so much for responding!

I’m far from an expert, so I can only hypothesise (although I kinda know what outcome I want). I think `n_intents` above lists intents **supported** by the profile ([Table 20, p. 19 of spec](https://color.org/specification/ICC.1-2001-04.pdf#page=28&zoom=140,-309,773)). I don’t think Photoshop writes a custom profile in this scenario (it does for some ancient dotgain custom values, but that’s unrelated). I think it writes the same ISOnewspaper26v4 profile and the supported intents are all the same. It’s just this other bit of profile metadata ([Table 18, p. 17 of spec](https://color.org/specification/ICC.1-2001-04.pdf#page=25&zoom=140,-309,185)) it modifies depending on the actual rendering intent used for the conversion **to** that profile. The spec for that tag says:
> Typically, the user or application will set the rendering intent dynamically at runtime or embedding time. Therefore, this flag may not have any meaning **until the profile is used in some context**, e.g in a DeviceLink or an embedded source profile.

(emphasis mine; here – it was used, so the tag was modified accordingly).

To check, I took those two images I attached above, extracted the profiles from them using ImageMagick and displayed them side-by-side in Colorsync Utility (I [attach them here too](https://github.com/libvips/libvips/files/11365470/Profiles.extracted.from.images.converted.in.Photoshop.RGBtoCMYK.using.different.Rendering.Intents.zip)). Notice how supported intents are the same (all AtoBs and BtoAs), but the “Rendering Intent” tag differs:

![image](https://user-images.githubusercontent.com/6032869/235455801-dc127fb1-db7c-423d-a945-76fc3f192c3f.png)

I also attach the [source RGB image](https://github.com/libvips/libvips/files/11365663/Source.RGB.image.zip) used for Photoshop conversions. My ideal outcome is to turn both CMYK conversions attached in first post back to RGB and match this image as closely as possible (for that to happen, the intents used must match). Using one command.

I’m currently struggling to use VIPS command line to make it display verbose output like in your snippets, but I don’t want this thread (which I hope has merit) to turn into a pre-school on command line. I will consult with my more knowledgable colleagues and come back with code examples.

Currently, when I try one VIPS command on both files
```
vips icc_transform in_toCMYK_using[Relative|Perceptual].jpg vips_out_[Rel|Perc].jpg sRGB.icm --embedded --black-point-compensation
```
I get different RGB results back. And the one from Perceptual looks off, so it must be using default Relative intent?

**Reply by jcupitt** [reply]
Ah you're right, I missed that bit of the lcms API. I now see:

```
$ vips icc_import ~/pics/cmyk2.jpg x.v --input-profile ~/pics/ISOnewspaper26v4.icc   
loading input profile, intent relative, from blob 0x563e07f300d0
icc profile loaded from blob to make: 0x563e07fc9c00
description: ISOnewspaper26v4
copyright: Copyright by LOGO GmbH, Steinfurt
intent: 0
profile class: 0x70727472
PCS: 0x4c616220 (Lab )
matrix shaper: 0
version: 2.4
n_intents = 10
  0: Perceptual, in CLUT = 1, out CLUT = 1
...
```

So yes it's been tagged as defaulting to perceptual.

I agree, I think libvips needs a new intent (perhaps `default`? or `profile-default`?) meaning "use the intent the profile suggests" would be a good idea. I'll make an enhancement issue.

In the meantime, I suppose you could use `exiftool` to get the default intent from an image. Would that work?

**Comment by jcupitt** [comment]
I made an enhancement issue for this feature: https://github.com/libvips/libvips/issues/3475

Thanks for the suggestion!

**Comment by paperboyo** [comment]
Thank you so much!
> perhaps default? or profile-default?

FWIW, I would make sure the name of the new `intent` option that it comes from the embedded (or supplied?) source image profile. So `embedded` would be clearest if it wouldn’t already be used, haha. `ProfileSupplied`? But I’m bad at naming.

> In the meantime, I suppose you could use exiftool to get the default intent from an image. Would that work?

Oh, yes, definitely.

Just checked IM and it doesn’t do what Photoshop does (modify this tag in profiles embedded in output images). Wondering if this is Adobe invention or IM’s omission or… Spec isn’t super-clear, to be honest. Maybe someone at lcms would know if this is something VIPS should do too? (but here I’m just trying to be a good citizen, huge majority of images I’m gonna deal with were/are converted by Photoshop).

Thanks again.
